Updated governance: Working Groups (WGs) (#14603)
PowerShell/PowerShell: docs/community/governance.md

* [**PowerShell Committee**](#powershell-committee): A committee of project owners who are responsible for design decisions,
  approving [RFCs][RFC-repo], and approving new maintainers/committee members
* [**Repository maintainer**](#repository-maintainers): An individual responsible for merging pull requests (PRs) into `master` when all requirements are met (code review, tests, docs, and RFC approval as applicable).
  Repository Maintainers are the only people with write permissions for the `master` branch.

> * [**Working Groups (WGs)**](#working-groups-(wgs)) are collections of people responsible for

Reviewer: NIT: you have slightly different language between this list of definitions and individual sections (line 95 below).  Prefer `contributors` to just `people` here as well.

For more information on Repository Maintainers--their responsibilities, who they are, and how one becomes a Maintainer--see the [README for Repository Maintainers][maintainers].

## Working Groups (WGs)

[Working Groups (WGs)][wg] are collections of contributors with knowledge of specific components or
technologies in the PowerShell domain.
They are responsible for issue triage/acceptance, code reviews, and providing their expertise to

> others in issues, PRs, and RFC discussions.

Author: I don't think so....milestone is still something that I think is going to happen organically based on how RFCs and prototypes are timed with releases

> others in issues, PRs, and RFC discussions.

Reviewer 2: If we will have some work groups and they will really work they need to set priorities - in the case custom/community/workgroup milestone labels can help.

They have [write access](https://help.github.com/articles/permission-levels-for-an-organization-repository/) to the PowerShell repository which gives them the power to:

1. `git push` to all branches *except* `master`.
1. Merge pull requests to all branches *except* `master` (though this should not be common given that [`master`is the only long-living branch](../git/README.md#understand-branches)).
1. Assign labels, milestones, and people to [issues](https://guides.github.com/features/issues/).

### Working Group Responsibilities

If you are a member of a Working Group, you are expected to be actively involved in any development, design, or contributions in the focus area of the WG.
More information on the responsibilities of Working Groups can be found [here][wg],

> while current WG definitions and membership can be found [here][wg-definitions].

Reviewer: Should we make clear that contributors don't have to be members of a WG to participate in a WG area, but WG members are empowered to make approval and rejection decisions

If you are a Working Group member:

1. **DO** triage and contribute to discussions in issues and PRs that have `WG-*` labels assigned
1. **DO** regularly communicate with other members of your WG to coordinate decisions about

>    whether issues should move forward

Reviewer 2: Since MSFT leads the project and control a compliance Teams is best choice. MSFT could create work spaces there for every WG.

>    whether issues should move forward

Reviewer 3: Teams is a good choice from that perspective, but in my experience if you already have a teams org you can't really operate Teams in multiple orgs at once. You have to log out from one to log into a different one, and that's not a useable workflow for any community members that work at a place that already uses Teams.

>    whether issues should move forward

Reviewer 2: > you can't really operate Teams in multiple orgs at once.

This is not so critical for the work. 
I don't strong vote for Teams but I believe it is good start.

>    whether issues should move forward

Reviewer 3: It seems the intention is to promote communication, so I think having some solid steps laid out for this aspect would go quite a long way.

>    whether issues should move forward

Author: By design, I didn't want to overly prescriptive about where comms should take place, especially as we experiment in this preliminary period to see what works best. 

However, I do think it's extremely important that WGs have some form of *private* communication. Something we learned in trying to record Committee meetings is that high-bandwidth discussions sometimes need to be "unfiltered" compared to what is expected in public issues and PRs. 

Another consideration is how timezone distribution across different WGs may affect whether folks want to hold regular voice/video calls vs. async text communications, etc. 

I will also +1 that Teams is probably not the right solution for anything folks external to the company. There are issues with Teams as a cross-organizational tool, especially if you're already using Teams as a business tool (i.e. acct switching is painful). As a counter-example, we actually use a Twitter DM group to discuss private matters within the Committee (a handful of whom are not Microsoft employees).

>    whether issues should move forward

Author: Also, we will offer additional recommendations as we learn from our own experiences operationalizing this process.

1. **DO** make decisions on whether or not issues should proceed forward with implementations or RFCs

> 1. **DO** assign the [correct labels][issue-process] to issues

Reviewer 4: that assumes we will have the privilege to do so. is that something we will have to enable?

Author: Naively, we should make WGs Teams within the PowerShell organization and give them Triage permissions within the PS org. However, we may not be able to add external collaborators without adding them to the PS org

Given the fullness of time:
* if we're can add external contributors to the PS org (this may not be something we can do, we'll go and check), we should rework permissions such that WG teams only have access to public repos
* we can also use a bot that enables trusted users to use comment-based commands to label on their behalf

In the intermediate, we can simply add external contributors directly to the PS repo with Triage permissions.

Reviewer 2: I think we need initially auto set `untriaged` label until WG makes a conclusion for the issue.

Author: Our thinking thus far has been for WG members to be the ones to add an `Issue - *` label as the "2nd order triage", but I kind of like this better. Will raise it up with the team

Reviewer 2: Related discussion with @daxian-dbw https://github.com/PowerShell/PowerShell/pull/14645#issuecomment-764853472

Reviewer 5: You can give triage without making the part of the org.  I have done it in PowerShell-Docker

Reviewer 3: Mhm. I have triage permissions. Dunno how it's done, but it's doable. Think it has to be done on an individual basis, though. For the purposes of WGs it might make more sense to have people be a member of the org under a specific WG "team"/group with appropriate permissions set. Less hassle, I would think.

> 1. **DO** assign yourself to issues labeled with your area of expertise when you plan to implement

Reviewer: I think we should have a bit stronger language here, people should assign issues to themselves (note that I think only Contributors have this permission or show up in the list) if they are ACTIVELY working on the issue.  We want to avoid people not working on issues because someone else self-assigned (or assigned to someone else) and no one is actually working on it.

>    them in a reasonable timeframe

Reviewer 4: ```suggestion
   a change in a reasonable timeframe
```

Author: ```suggestion
   the proposal in a reasonable timeframe
```

1. **DO** code reviews for PRs where you're assigned or in your WG
   (while reviewing PRs, leave your comment even if everything looks good - a simple "Looks good to me" or "LGTM" will suffice, so that we know someone has already taken a look at it).
1. **DO** make sure contributors are following the [contributor guidelines](../../.github/CONTRIBUTING.md)
1. **DO** ensure that contributors [write Pester tests][pester] for all new/changed functionality
1. **DO** ensure that contributors [write documentation][docs-contributing] for all new-/changed functionality

> 1. **DO** encourage contributors to refer to issues in their pull request description (e.g. `Resolves issue #123`)

Reviewer 4: hmmm, i wonder if this should be more than encouragement?

Author: ```suggestion
1. **DO** ensure that contributors refer to issues in their pull request description (e.g. `Resolves issue #123`)
```

> 1. **DO** ensure that contributors create meaningful titles for all PRs, editing their title if necessary

Reviewer: Perhaps a bit more explanation that PR titles end up in the changelog, so titles should be user facing.

> 1. **DO** encourage contributors to create meaningful titles for all PRs, editing their title if necessary

Author: ```suggestion
1. **DO** ensure that contributors create meaningful titles for all PRs, editing their title if necessary
```
